### PR TITLE
Don't insert imports when method calls are created through searcher

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
When method call is inserted as a suggestion from searcher, the import of the relevant module won't be added anymore.
It is not needed and may cause trouble with name collisions, like #1145.
Issue #1178

### Checklist
Please include the following checklist in your PR:

- [x] The `CHANGELOG.md` was updated with the changes introduced in this PR.
- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
- [x] All code has automatic tests where possible.
- [x] All code has been profiled where possible.
- [x] All code has been manually tested in the IDE.
- [x] All code has been manually tested in the "debug/interface" scene.
- [x] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [x] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

### CODE REVIEW

**Review by farmaazon — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,8 @@ read the notes of the `Enso 2.0.0-alpha.1` release.
 - [Assigning intermediate expressions to values][1067]. Nodes added with searcher will have their 
   values automatically assigned to a newly generated variables. This allows Enso Engine to cache
   intermediate values, improving visualization performance.
+- [Not adding spurious imports][1209]. Fixed case where IDE would generate spurious import 
+  declarations when inserting method calls.
```
Comment (wdanilo) [inline CHANGELOG.md]: Could you describe the issue a little bit more using simpler words and maybe bigger description? The changelog is for users snd they should understand what it is about. Even "inserting method calls" is not really understandable fully, as in the GUI its not obvious to every user what that's mean.

I would go this way (written fast, contains errors, just a scratch description):
```
[Not adding unnecessary library imports][1209]. Fixed situation when IDE was adding unnecessary libraries imports when selecting hints from node searcher. This makes the generated textual code easier to read.
```

**Review by wdanilo — APPROVED** [review]

**Review by farmaazon — CHANGES_REQUESTED** [review]
Unfortunately, the reality is more complex than I expected. Actually there is scenario which works wrong with the current solution: press tab without any node selected, pick Table.new and press enter.

the new is a method, but it is a method of not imported module "Table". So actually we should skip importing only when we pick a method for this argument.

**Comment by kustosz** [comment]
@farmaazon @mwu-tow it's even more complex! Even if you have the `this`, some methods are extension methods. For an example, `Json.to_table` is a method in defined `Table`, even though `Json` comes from `Base`. So the rule for "import is not necessary" is:
1. The method being called comes from the same module as `this`. In this case, never import.
2. The method being called comes from a different module than `this`. Then check if the method's module is already imported and don't import if it aready is.

**Comment by farmaazon** [comment]
As for point 2: in scope of this task is only checking if we literally have the same import (as we do with other suggestions).

**Comment by kustosz** [comment]
@farmaazon sure, point 1 is the most important one, because it could lead you towards overoptimization – i.e. not importing extension methods, which would break programs (and not just make them ugly by too many imports)

**Comment by kustosz** [comment]
I have not read this code btw. it's very possible you've already taken everything into account :P Just saw this discussion and randomly commented

**Comment by farmaazon** [comment]
No, no, the point 1 in your comment is important (cc @mwu), and no, we don't handle it in our code.

File: src/rust/ide/src/model/suggestion_database/entry.rs [hunk 2]
```diff
@@ -103,25 +114,45 @@ impl Entry {
...
+        let code = match this_expr {
+            Some(this_expr) => format!("{}{}{}",this_expr,ast::opr::predefined::ACCESS,self.name),
+            None            => self.name.clone(),
+        };
+
+        CodeToInsert {code,imports}
+    }
+
+    /// Check if this is a non-extension module method.
+    // TODO [mwu] Currently we have no means to recognize module extension methods.
```
Comment (farmaazon) [inline src/rust/ide/src/model/suggestion_database/entry.rs]: Could you add an issue for TODOs here? And add a note, that module extension methods will be treated as non-module methods.

File: src/rust/ide/src/model/suggestion_database/entry.rs [hunk 3]
```diff
@@ -103,25 +114,45 @@ impl Entry {
...
-            self.name.clone()
+    /// Describes code insertion to be done when Searcher input suggestion is picked.
+    pub fn code_to_insert
+    (&self, current_module:Option<&module::QualifiedName>, generate_this:bool) -> CodeToInsert {
+        let mut imports = BTreeSet::new();
+
+        // Regular methods are found through this-lookup. Other entities need to be visible.
+        // Import if not already defined in this module.
+        if !self.is_regular_method() && !current_module.contains(&&self.module) {
+            imports.insert(self.module.clone());
```
Comment (farmaazon) [inline src/rust/ide/src/model/suggestion_database/entry.rs]: I don't understand the comment.

File: src/rust/ide/src/model/suggestion_database/entry.rs [hunk 4]
```diff
@@ -164,6 +195,20 @@ impl Entry {
     pub fn invocation_info(&self) -> span_tree::generate::context::CalledMethodInfo {
         self.into()
     }
+
+    /// Check if this is a regular method (i.e. non-extension method).
+    pub fn is_regular_method(&self) -> bool {
+        println!("{:?}",self);
```
Comment (farmaazon) [inline src/rust/ide/src/model/suggestion_database/entry.rs]: println left.

**Review by farmaazon — COMMENTED** [review]
Please, update the branch so I could easily test it with all new changes.

**Review by farmaazon — APPROVED** [review]

**Pull request merged**